discovery-provider/src/tasks/index.py

```diff
@@ -254,13 +256,106 @@ def fetch_tx_receipts(self, block):
+            resp_json.get(parameter) if resp_json.get(parameter) != None else value
+        )
+    return metadata
+
+
+async def fetch_metadata_from_gateway_endpoints(
+    cid_metadata,
+    cids_txhash_set,
+    cid_to_user_id,
+    user_to_replica_set,
+    cnode_endpoints,
+    cid_type,
+    block_hash,
+    block_number,
+    fetch_from_replica_set=True,
+):
+    async with aiohttp.ClientSession() as async_session:
+        futures = []
+        cid_futures_map = {}
+        cid_txhash_map = {cid: txhash for cid, txhash in cids_txhash_set}
+
+        for cid in cid_txhash_map:
+            if cid in cid_metadata:
+                continue  # already fetched
+            user_id = cid_to_user_id[cid]
+
+            gateway_endpoints = cnode_endpoints
+            if fetch_from_replica_set and user_id and user_id in user_to_replica_set:
+                user_replica_set = user_to_replica_set[
+                    user_id
+                ]  # user or track owner's replica set
+                if not user_replica_set:
+                    continue  # try all gateway endpoints on later attempt
+                gateway_endpoints = user_replica_set.split(",")
+
+            logger.info(f"isaac fetching CID {cid} from {gateway_endpoints}")
+            for gateway_endpoint in gateway_endpoints:
+                future = asyncio.ensure_future(
+                    update_task.ipfs_client.get_metadata_async(
+                        async_session, cid, gateway_endpoint
+                    )
+                )
+                if cid not in cid_futures_map:
+                    cid_futures_map[cid] = set()
+                cid_futures_map[cid].add(future)
+                futures.append(future)
+
+        try:
+            for future in asyncio.as_completed(futures, timeout=5):
+                try:
+                    future_result = await future
+                except asyncio.CancelledError:
+                    pass  # swallow canceled requests
+
+                if not future_result:
+                    continue
+
+                cid, metadata_json = future_result
+
+                metadata_format = (
+                    track_metadata_format
+                    if cid_type[cid] == "track"
+                    else user_metadata_format
+                )
+
+                formatted_json = get_metadata_from_json(metadata_format, metadata_json)
+
+                if formatted_json != metadata_format:
+                    cid_metadata[cid] = formatted_json
+
+                    if len(cid_metadata) == len(cid_txhash_map):
+                        break  # fetched all metadata
+
+                    for other_future in cid_futures_map[cid]:
+                        if other_future == future:
+                            continue
+                        other_future.cancel()  # cancel other pending requests
+        except asyncio.TimeoutError:
+            logger.info("index.py | fetch_cid_metadata TimeoutError")
```

Comment:
this doesn't raise an error the first time because it retries the second time with the entire list of gateways?

Reply:
yes